### PR TITLE
matejovic

### PR DESCRIPTION
**Pull request by chuchulo**
I would add more unit tests

### CODE REVIEW

**Comment by marcelmojzis** [comment]
@chuchulo Could you please provide at least few tests?

File: src/main/scala/com/gwi/Job.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            context.self ! ProcessLine
+            processing(state.copy(status = Running))
+          } else Behaviors.same
+        case Cancel =>
+          if (state.status == Running || state.status == Scheduled) {
+            processing(state.copy(status = Canceled, endTime = Some(System.currentTimeMillis())))
+          } else Behaviors.same
+        case ProcessLine =>
+          if (state.status == Running && state.iterator.hasNext) {
+            val line = state.iterator.next
```
Comment (marcelmojzis) [inline src/main/scala/com/gwi/Job.scala]: Could you elaborate on what happens when connection becomes very slow?

File: src/main/scala/com/gwi/Job.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    startTime: Long,
+    endTime: Option[Long],
+    status: Status,
+    taskRepository: ActorRef[TaskRepository.Command]
+  )
+  def apply(id: TaskRepository.Id, sourceUri: Uri, taskRepository: ActorRef[TaskRepository.Command]) = {
+    val start     = System.currentTimeMillis()
+    val targetUri = Uri(s"/tmp/$id.json")
+    val bw        = new BufferedWriter(new FileWriter(new File(targetUri.toString()),true))
+    val source    = Source.fromURL(sourceUri.toString())
```
Comment (marcelmojzis) [inline src/main/scala/com/gwi/Job.scala:47]: How would you handle failures?

File: src/main/scala/com/gwi/Job.scala (hunk 1 above)
Comment (chuchulo) [reply src/main/scala/com/gwi/Job.scala]: what do you mean by connection?

Comment (chuchulo) [reply src/main/scala/com/gwi/Job.scala]: what particular case do you have in your mind?

File: src/main/scala/com/gwi/Playground.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.HttpMethods._
+import akka.http.scaladsl.model._
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.Sink
+import scala.concurrent.Future
+
+object Playground extends App {
```
Comment (chuchulo) [inline src/main/scala/com/gwi/Playground.scala]: sorry for this. This should be removed :)

File: src/main/scala/com/gwi/Job.scala (hunk 1 above)
Comment (marcelmojzis) [reply src/main/scala/com/gwi/Job.scala]: The iterator reads data from the underlying URLConnection. The connection might fail, become stale, very slow...

Comment (chuchulo) [reply src/main/scala/com/gwi/Job.scala]: if the connection fails, the IOException would be thrown, I implemented some error handling, If the connection becomes slow, Some timeout for job would be needed to implement

File: src/main/scala/com/gwi/Job.scala (hunk 2 above)
Comment (chuchulo) [reply src/main/scala/com/gwi/Job.scala:47]: I implemented supervision.restart strategy and trying to set proper status for the job when the IOException is being thrown

**Review by chuchulo — COMMENTED** [review]
omg 🤦  I am so sorry I haven' realiZed I have to submit my comments at the end once again.

File: src/main/scala/com/gwi/Job.scala (hunk 1 above)
Comment (marcelmojzis) [reply src/main/scala/com/gwi/Job.scala]: If connection becomes slow ,`state.iterator.next` will block the calling thread and will have negative impact on actor dispatchers...

**Pull request closed without merging**